nmigen/hdl/ast.py

```diff
@@ -1186,7 +1186,13 @@ def lower(self):
 
     def _lazy_lower(self):
         if self.__lowered is None:
-            self.__lowered = Value.cast(self.lower())
+            lowered = self.lower()
+            if isinstance(lowered, UserValue):
+                lowered = lowered._lazy_lower()
+            elif not isinstance(lowered, Value):
+                raise ValueError("UserValue.lower() needs to return a nMigen Value or UserValue, found {}"
```

Comment:
Nit: our convention is to show a repr in error messages such as this, which should provide more information.
```suggestion
                raise ValueError("UserValue.lower() needs to return a nMigen Value or UserValue, not {!r}"
                                                 .format(lowered))
```

Reply:
ok this check was stupid anyway since it doesnt allow for returning e.g. integers and all other cases are cought by Value.cast()

Reply:
Good catch.